src/index.mjs

```diff
@@ -35,85 +35,82 @@ const RETURN = 13;
 const SPACE = 32;
 const SLASH = 47;
 
-const MODE_WHITESPACE = 0;
-const MODE_TEXT = 1;
+const MODE_TEXT = 0;
+const MODE_WHITESPACE = 1;
 const MODE_TAGNAME = 9;
 const MODE_ATTRIBUTE = 13;
 const MODE_SKIP = 47;
 
 /** Create a template function given strings from a tagged template. */
 const build = (statics) => {
-	let mode = MODE_WHITESPACE;
+	let mode = MODE_TEXT;
 	let out = 'return ';
 	let buffer = '';
-	let field = '';
-	let hasChildren = 0;
+	let tagClose = '';
+	let childClose = '';
 	let props = '';
 	let propsClose = '';
 	let spreadClose = '';
 	let quote = 0;
-	let slash, charCode, inTag, propName, propHasValue;
+	let fallbackPropValue = true;
+	let charCode, propName;
 
-	const commit = () => {
-		if (!inTag) {
+	const commit = field => {
+		if (mode === MODE_TEXT) {
 			if (field || (buffer = buffer.replace(/^\s*\n\s*|\s*\n\s*$/g,''))) {
-				if (hasChildren++) out += ',';
-				out += field || stringify(buffer);
+				out += childClose + (field || stringify(buffer));
+				childClose = ',';
 			}
 		}
 		else if (mode === MODE_TAGNAME) {
-			if (hasChildren++) out += ',';
-			out += 'h(' + (field || stringify(buffer));
+			out += childClose + 'h(' + (field || stringify(buffer));
+			childClose = ',';
 			mode = MODE_WHITESPACE;
 		}
-		else if (mode === MODE_ATTRIBUTE || (mode === MODE_WHITESPACE && buffer === '...')) {
+		else if (mode === MODE_WHITESPACE && buffer === '...') {
+			if (!spreadClose) {
+				props = 'Object.assign(' + (props || '{}');
+			}
+			props += propsClose + ',' + (field || '');
+			propsClose = '';
+			spreadClose = ')';
+		}
+		else if (mode === MODE_ATTRIBUTE || mode === MODE_WHITESPACE) {
 			if (mode === MODE_WHITESPACE) {
-				if (!spreadClose) {
-					spreadClose = ')';
-					if (!props) props = 'Object.assign({}';
-					else props = 'Object.assign(' + props;
-				}
-				props += propsClose + ',' + field;
-				propsClose = '';
+				propName = buffer;
+				buffer = '';
 			}
-			else if (propName) {
-				if (!props) props += '{';
-				else props += ',' + (propsClose ? '' : '{');
+			
+			if (propName) {
+				if (!props) {
+					props += '{';
+				}
+				else {
+					props += ',' + (propsClose ? '' : '{');
+				}
 				propsClose = '}';
```

Comment:
I think if you move this line under the `props +=` below, terser will generate smaller output, right now it does: `l+=l?","+(g?"":"{"):"{",g="}",l+=e(r)` when it could be `l+=l?","+(g?"":"{"):"{"+e(r),g="}"` which should shave 3 more bytes maybe?

Reply:
That is, the whole block above could be:
```
props += (props ? ',' + (propsClose ? '' : '{') : '{') + stringify(propName) + ':' (field || stringify(buffer || fallbackPropValue));
```
which terser May be smart enough to do if you move this line down!

Reply:
Thanks! I tried different variations of this, and hand-rolling the `props +=` line into one seemed to make the *compressed* files smallest. However, combined with another similar trick there was a significant effect:

&nbsp;| Before changes | After your change | After other change | After (your + other) change
-------------|----------------|------------------|----------------|-----------------------
htm.mjs.gz   | 603 B          | 602 B            | 602 B          | 600 B
htm.mjs.br   | 544 B          | 546 B            | 550 B          | 537 B
htm.umd.js.gz| 669 B          | 668 B            | 668 B          | 667 B
htm.umd.js.br| 595 B          | 597 B            | 593 B          | 592 B

I'll push a commit for this.